runtime/src/impls.rs

```diff
@@ -17,24 +17,66 @@
 //! Some configurable implementations as associated type for the substrate runtime.
 
 use crate::constants::fee::TARGET_BLOCK_FULLNESS;
-use crate::{Call, MaximumBlockWeight, Runtime};
+use crate::{Call, MaximumBlockWeight, NegativeImbalance, PositiveImbalance, Runtime};
 use cennznet_primitives::{
 	traits::{BuyFeeAsset, IsGasMeteredCall},
 	types::{Balance, FeeExchange},
 };
 use crml_transaction_payment::GAS_FEE_EXCHANGE_KEY;
 use frame_support::{
+	additional_traits::{InherentAssetIdProvider, MultiCurrencyAccounting},
 	storage,
-	traits::{Currency, ExistenceRequirement, Get, OnUnbalanced, WithdrawReason},
+	traits::{Currency, ExistenceRequirement, Get, Imbalance, OnUnbalanced, WithdrawReason},
 	weights::Weight,
 };
 use pallet_contracts::{Gas, GasMeter};
 use pallet_generic_asset::StakingAssetCurrency;
+use pallet_staking::RewardDestination;
 use sp_runtime::{
 	traits::{CheckedMul, CheckedSub, Convert, SaturatedConversion, Saturating, UniqueSaturatedFrom, Zero},
 	DispatchError, Fixed64,
 };
 
+type CennzxSpot<T> = crml_cennzx_spot::Module<T>;
+type Contracts<T> = pallet_contracts::Module<T>;
+type GenericAsset<T> = pallet_generic_asset::Module<T>;
+type Staking<T> = pallet_staking::Module<T>;
+
+pub struct SplitToAllValidators;
+
+/// This handles the ```NegativeImbalance``` created for transaction fee.
+/// The reward is split evenly and distributed to all of the current elected validators.
+/// The remainder from the division are burned.
+impl OnUnbalanced<NegativeImbalance> for SplitToAllValidators {
+	fn on_nonzero_unbalanced(imbalance: NegativeImbalance) {
+		let validators = Staking::<Runtime>::current_elected();
```

Comment:
Can we be certain that on_nonzero_unbalanced would be called before a new set of validators are elected for a new era? In other words, are we rewarding the correct set of validators? [It's just a question]

Reply:
this function is only called during validate(...) function, which I think is used to verify transactions. Therefore it will only be called when we are processing transactions - i.e. during an session when validators are all set and present.
Worse case scenario we have no validators and we just burn the transaction fee

Reply:
Also since the chain is running single threaded, I think it would not be possible for the validators to change in the middle of a transaction.
In addition, we make a copy of the list of validators, therefore even if the validators change, we would be OK